File: ml-benchmark/autolifter/single-pass/is_sorted.f
Config ExtraGrammar = "AutoLifter";

Inductive List = cons {Int, List} | nil Unit;

single_pass = \v: List -> Bool.
  let run = fix (
    \f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {h, t} -> cons {h, f t}
    end
  ) in \xs: List.
  v (run xs);

inf = 100;

/*User provided programs*/

is_sorted = fix (
  \f: Int -> List -> Bool. \pre: Int. \xs: List.
  match xs with
    nil _ -> true
  | cons {h, t} -> if >= pre h then false else f h t
  end
) (- 0 inf);

main = single_pass is_sorted;

File: ml-benchmark/autolifter/single-pass/is_sorted.ml
type list = Nil of unit | Cons of int * list

let single_pass v =
  let rec run xs =
    match xs with
    | Nil _ -> xs
    | Cons (h, t) ->
        Cons (h, run t)
  in
  fun xs -> v (run xs)

let inf = 100

let rec is_sorted_f pre xs =
  match xs with
  | Nil _ -> true
  | Cons (h, t) ->
      if pre >= h then false else is_sorted_f h t

let is_sorted xs =
  is_sorted_f (0 - inf) xs

let program = single_pass is_sorted